### PR TITLE
Add Target column to the LinkFieldIndex table

### LINKED ISSUE #16378: Rebuilding Lucene Index fails
### Describe the bug
Rebuilding Lucene Index in one of my tenants fails with the Error `SqlException: Invalid column name 'Target'.` and the Stacktrace shown below. Creating a fresh Lucene Index works on that tenant, rebuilding it fails.

On a different tenant rebuilding works fine. But the other tenant is quite young compared to the one that creates the problems. 

My guess would be a Migration problem of the DataBase, as the problem seem to be the way to access content items:
`OrchardCore.ContentManagement.DefaultContentManager.GetAsync(IEnumerable<string> contentItemIds, VersionOptions options)`

### Orchard Core version

2.0.0.-preview-18254

### To Reproduce
Seems to be dependent on the tenant/db, so likely not reproducible. 

1. Go to Admin Backend 
3. Ensure these Features are enabled: 
`Lucene` (`Content Types` `Contents` `Indexing` `Liquid` `Settings`), `Search`, `PDF Media Indexing` (`Media Indexing`)
4. Go to Search->Indexing->Lucene Indices
5. Ensure you have a defined Search Index (and some content you can search for)
6. Select your Lucene Index and click "Rebuild" or "Reset"
7. See error
8. All searches on frontend lead to empty result lists now.

### Expected behavior
Should Rebuild or Reset the index

### Logs and screenshots
```
SqlException: Invalid column name 'Target'.
System.Threading.Tasks.ContinuationResultTaskFromResultTask<TAntecedentResult, TResult>.InnerInvoke()
System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, object state)
System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, object state)
System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref Task currentTaskSlot, Thread threadPoolThread)
YesSql.Commands.BatchCommand.ExecuteAsync(DbConnection connection, DbTransaction transaction, ISqlDialect dialect, ILogger logger)
YesSql.Session.FlushInternalAsync(bool saving)
YesSql.Session.FlushInternalAsync(bool saving)
YesSql.Services.DefaultQuery+Query<T>.ListImpl()
OrchardCore.ContentManagement.DefaultContentManager.GetAsync(IEnumerable<string> contentItemIds, VersionOptions options)
OrchardCore.Search.Lucene.LuceneIndexingService+<>c__DisplayClass10_0+<<ProcessContentItemsAsync>b__0>d.MoveNext()
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Search.Lucene.LuceneIndexingService.ProcessContentItemsAsync(string indexName)
OrchardCore.Search.Lucene.Controllers.AdminController.Rebuild(string id)
lambda_method2041(Closure , object )
Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+TaskOfActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask<IActionResult> actionResultValueTask)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
OrchardCore.Seo.Services.RobotsMiddleware.Invoke(HttpContext httpContext)
SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext httpContext, bool retry)
OrchardCore.Liquid.ScriptsMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
OrchardCore.Diagnostics.DiagnosticsStartupFilter+<>c__DisplayClass3_0+<<Configure>b__1>d.MoveNext()
Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
OrchardCore.ContentPreview.PreviewStartupFilter+<>c+<<Configure>b__0_1>d.MoveNext()
OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext)
OrchardCore.Modules.ModularTenantContainerMiddleware+<>c__DisplayClass4_0+<<Invoke>b__0>d.MoveNext()
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute, bool activateShell)
OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #16378

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs [hunk 1]
```diff
@@ -807,5 +807,14 @@ public async Task<int> UpdateFrom4Async()
 
             return 5;
         }
+
+        public async Task<int> UpdateFrom5Async()
+        {
+            await SchemaBuilder.AlterIndexTableAsync<LinkFieldIndex>(table => table
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs:813]: This is needed in `CreateAsync` too.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs [hunk 2]
```diff
@@ -89,11 +90,11 @@ public override void Describe(DescribeContext<ContentItem> context)
                                 ContentType = contentItem.ContentType,
                                 ContentPart = pair.Definition.ContentTypePartDefinition.Name,
                                 ContentField = pair.Definition.Name,
-                                Url = pair.Field.Url?[..Math.Min(pair.Field.Url.Length, LinkFieldIndex.MaxUrlSize)],
+                                Url = pair.Field.Url?.Substring(0, Math.Min(pair.Field.Url.Length, LinkFieldIndex.MaxUrlSize)),
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs:93]: Why go back to `Substring()`?

**Review by Piedone — CHANGES_REQUESTED** [review]
I only briefly checked out the code, but I'll be able to properly test and asses it once I'm back from holiday, next Tuesday.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs:813]: Its not needed. `CreateAsync` returns 3. From there we do other steps including the new step. If we add the column during create, then this step will through an exception

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs (hunk 2 above)
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs:93]: While I agree, there are many cases of it in the source code, and just 5 for `Substring()`:

image

(And 3 more for another overload.)

At one point, the new syntax was adopted, so I'd prefer consistency. We can keep this now, but we'll need to have a formal recommendations once we enforce either of these with analyzers.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs (hunk 1 above)
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs:813]: [`CreateAsync` returns 5](https://github.com/OrchardCMS/OrchardCore/blob/8e8f2d78ddbe5106d9fcd0f371ffb28c2cf116f4/src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs#L437) for the same reason.

So, we need to add the column there too, and update its return value to `6`.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs (hunk 2 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/LinkFieldIndexProvider.cs:93]: At some point we changed that preference in the `.editorconfig` file so even the IDE no longer suggest to use this syntax. This was discussed one time in a meeting and the suggestion was to use a readable code over the new fancy syntax. I think this suggestion came from @sebastienros

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs:814]: I don't know why I saw it return 3 not return 5 :) I changed it.

**Comment by sebastienros** [comment]
So nobody agrees with my comments ? (https://github.com/OrchardCMS/OrchardCore/pull/16326#issuecomment-2191143589) we just accept anything in the index? (or graphql, sql index, ... why stop there). Even when there is an obvious alternative: adding fields using existing types.

**Comment by MikeAlhayek** [comment]
closing this in favor of #16397

**Pull request closed without merging**